### PR TITLE
Promote service load balancer finalizer to Beta

### LINKED ISSUE #53451: service controller can leave orphaned load balancers after services are deleted
/kind bug
/area cloudprovider

**Current**

The [service controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L735) fetches the service from the apiserver to sync. If the service no longer exists, then the load balancer is deleted. If the deletion fails (say networking issues or something), then the service key will be added back to the queue to be retried later. This works well until the service controller is restarted while a delete is in process or queued to be processed. At that point, the new service controller that takes over will never see the deleted service since the service has been deleted from etcd. 

**Proposal**

The service account controller should be a [finalizer](https://github.com/kubernetes/community/blob/58b1c30d95719749068497ba35dfe4c64b21aa72/contributors/design-proposals/api-machinery/garbage-collection.md#orphaning-the-descendants-with-orphan-finalizer) on all services. Then we can ensure that the load balancer is successfully deleted prior to the service being deleted. That will make the delete process resilient to restarts / changes in leadership. 

cc @andrewsykim @wlan0 @luxas @prydie

### PR DESCRIPTION
**Pull request by MrHohn**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Ref https://github.com/kubernetes/enhancements/issues/980 and https://github.com/kubernetes/cloud-provider/issues/16. This PR promotes service load balancer finalizer to Beta and enable it by default. The e2e test has been patched to ensure finalizer is attached upon cluster upgrade.

The finalizer removal logic has been enabled in k8s 1.15 to allow a functioning cluster rollback.

**Which issue(s) this PR fixes**:

Fixes #53451

**Special notes for your reviewer**:
/assign @bowei @andrewsykim 

**Does this PR introduce a user-facing change?**:

```release-note
Finalizer Protection for Service LoadBalancers is now in Beta (enabled by default). This feature ensures the Service resource is not fully deleted until the correlating load balancer resources are deleted.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/e2e/framework/service/wait.go [hunk 1]
```diff
@@ -63,3 +67,49 @@ func WaitForServiceResponding(c clientset.Interface, ns, name string) error {
 		return true, nil
 	})
 }
+
+func WaitForServiceDeletedWithFinalizer(cs clientset.Interface, namespace, name string) {
```
Comment (MrHohn) [inline test/e2e/framework/service/wait.go]: These two functions are moved without modification to allow reusing --- the only change is replacing the hardcoded "service.kubernetes.io/load-balancer-cleanup" with the imported one.

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -1916,39 +1915,13 @@ var _ = SIGDescribe("Services", func() {
...
-		}
-	})
-
 	// This test verifies if service load balancer cleanup finalizer is properly
 	// handled during service lifecycle.
 	// 1. Create service with type=LoadBalancer. Finalizer should be added.
 	// 2. Update service to type=ClusterIP. Finalizer should be removed.
 	// 3. Update service to type=LoadBalancer. Finalizer should be added.
 	// 4. Delete service with type=LoadBalancer. Finalizer should be removed.
-	ginkgo.It("should handle load balancer cleanup finalizer for service [Slow] [Feature:ServiceFinalizer]", func() {
```
Comment (MrHohn) [inline test/e2e/network/service.go]: Keyword "[Feature:ServiceFinalizer]" has been removed to allow this to be run with the usual slow test suite.

File: test/e2e/network/service.go [hunk 3]
```diff
@@ -1916,39 +1915,13 @@ var _ = SIGDescribe("Services", func() {
 		}
 	})
 
-	// This test verifies if service load balancer cleanup finalizer can be removed
-	// when feature gate isn't enabled on the cluster.
-	// This ensures downgrading from higher version cluster will not break LoadBalancer
-	// type service.
-	ginkgo.It("should remove load balancer cleanup finalizer when service is deleted [Slow]", func() {
```
Comment (MrHohn) [inline test/e2e/network/service.go]: With finalizer now being enabled by default, this test case is no longer necessary.

**Comment by MrHohn** [comment]
Thanks for the quick review. Adding Tim for approval on the feature gate.
/assign @thockin

**Review by thockin — COMMENTED** [review]
Thanks!

/approve

**Pull request merged**